File: app/src/main/java/com/fishweeb/practical/ScoresState.java
package com.fishweeb.practical;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.SurfaceView;

import javax.crypto.spec.DHGenParameterSpec;

public class ScoresState implements StateBase
{
    private Bitmap background = null;
    private float bubbleSpawnTime = 2;
    private TextEntity headerText;
    private BackEntity backButton;
    private Paint HighScorePaint = new Paint();
    private Rect HighScoreTextBounds = new Rect();
    private Bitmap Coinbmp;
    private Paint moneyPaint = new Paint();
    private Rect moneyTextBounds = new Rect();

    @Override
    public String GetName() {
        return "Scores";
    }

    @Override
    public void OnEnter(SurfaceView _view)
    {
        background = Bitmap.createScaledBitmap(ImageManager.Instance.GetImage(IMAGE.I_MENUBACKGROUND),(int)GameSystem.Instance.GetScreenScale().x,(int)GameSystem.Instance.GetScreenScale().y,true);
        headerText = TextEntity.Create(GameSystem.Instance.ScreenScale.x*0.5f,0,"Scores",100);
        headerText.SetPosY(headerText.GetHeight()*0.5f);
        headerText.SetHeader(true);
        backButton = BackEntity.Create(0,0,headerText.height*1.3f,headerText.height*1.3f,"MainMenu");
        backButton.SetPosX(backButton.width*0.5f);
        backButton.SetPosY(backButton.height*0.5f);

        HighScorePaint.setColor(Color.BLACK);
        HighScorePaint.setTextSize(50);

        Coinbmp = BitmapFactory.decodeResource(_view.getResources(), R.drawable.coin);
        moneyPaint.setColor(Color.YELLOW);
        moneyPaint.setAlpha(200);
        moneyPaint.setTextSize(50);
        moneyPaint.setTextAlign(Paint.Align.LEFT);
        moneyPaint.setFakeBoldText(true);
    }

    @Override
    public void OnExit()
    {
        EntityManager.Instance.EmptyEntityList();
    }

    @Override
    public void Render(Canvas _canvas)
    {
        // Render anything
        _canvas.drawBitmap(background,0,0,null);
        ParticleManager.Instance.Render(_canvas);
        EntityManager.Instance.Render(_canvas);
        int count = 0;
        for (int i = 0;i<5;++i)
        {
            if (i >= GameSystem.Instance.GetHighScores().size())
            {
                _canvas.drawText((i+1) + ".---------",GameSystem.Instance.ScreenScale.x*0.1f,GameSystem.Instance.GetScreenScale().y*0.2f + HighScoreTextBounds.height()*1.2f*count,HighScorePaint);
                String scoreText = "0";
                HighScorePaint.getTextBounds(scoreText,0,scoreText.length(),HighScoreTextBounds);
                _canvas.drawText(scoreText,GameSystem.Instance.ScreenScale.x*0.9f - HighScoreTextBounds.width(),GameSystem.Instance.GetScreenScale().y*0.2f + HighScoreTextBounds.height()*1.2f*count,HighScorePaint);
                ++count;
                continue;
            }


            _canvas.drawText((i+1) + "." + GameSystem.Instance.GetHighScores().get(i).GetName(),GameSystem.Instance.ScreenScale.x*0.1f,GameSystem.Instance.GetScreenScale().y*0.2f + HighScoreTextBounds.height()*1.2f*count,HighScorePaint);
            String depthText = "" + (int)GameSystem.Instance.GetHighScores().get(i).GetScore();
            HighScorePaint.getTextBounds(depthText,0,depthText.length(),HighScoreTextBounds);
            _canvas.drawText(depthText,GameSystem.Instance.ScreenScale.x*0.9f - HighScoreTextBounds.width(),GameSystem.Instance.GetScreenScale().y*0.2f + HighScoreTextBounds.height()*1.2f*count,HighScorePaint);
            ++count;
        }

        //Money
        String moneyText = "" + PlayerInfo.Instance.GetMoney();
        moneyPaint.getTextBounds(moneyText,0,moneyText.length(),moneyTextBounds);
        _canvas.drawBitmap(Coinbmp,Coinbmp.getWidth()*0.5f,Coinbmp.getHeight()*0.5f + headerText.height*1.3f,null);
        _canvas.drawText(moneyText,Coinbmp.getWidth() + Coinbmp.getWidth()*0.7f,Coinbmp.getHeight() + headerText.height*1.3f + moneyTextBounds.height()*0.5f,moneyPaint);
        //Money
    }

    @Override
    public void Update(float _dt)
    {
        if (TouchManager.Instance.HasTouch())
        {
            for (int i = 0;i<3;++i)
            {
                ParticleManager.Instance.CreateParticle(PlayerInfo.Instance.GetEffectType(),TouchManager.Instance.GetPosX(),TouchManager.Instance.GetPosY());
            }
        }

        GameSystem.Instance.m_bubbleTimer += _dt;
        if (GameSystem.Instance.m_bubbleTimer >= bubbleSpawnTime)
        {
            ParticleObject newParticle = ParticleManager.Instance.FetchParticle(PARTICLETYPE.P_BUBBLE);
            int diameter = (int)(Math.random()*50 + 15);
            newParticle.width = diameter;
            newParticle.height = diameter;
            newParticle.position.x = ((float)Math.random() * (GameSystem.Instance.GetScreenScale().x - newParticle.width)) + (newParticle.width*0.5f);
            newParticle.position.y = GameSystem.Instance.GetScreenScale().y + newParticle.height;
            newParticle.velocity.x = (float)Math.random()*40-20;
            newParticle.velocity.y = -((float)Math.random()*160+80);
            newParticle.SetBMP(Bitmap.createScaledBitmap(ImageManager.Instance.GetImage(IMAGE.I_BUBBLE),diameter,diameter,true));
            GameSystem.Instance.m_bubbleTimer = 0;
        }
        ParticleManager.Instance.Update(_dt);
        EntityManager.Instance.Update(_dt);
    }
}

File: app/src/main/java/edu/androidgroupproject/ScoresState.kt
package com.fishweeb.practical

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.SurfaceView
import edu.androidgroupproject.R

class ScoresState : StateBase {
    private var background: Bitmap? = null
    private val bubbleSpawnTime = 2f
    private var headerText: TextEntity? = null
    private var backButton: BackEntity? = null
    private val HighScorePaint = Paint()
    private val HighScoreTextBounds = Rect()
    private var Coinbmp: Bitmap? = null
    private val moneyPaint = Paint()
    private val moneyTextBounds = Rect()

    override fun GetName(): String {
        return "Scores"
    }

    override fun OnEnter(_view: SurfaceView) {
        background = Bitmap.createScaledBitmap(
            ImageManager.Companion.Instance.GetImage(IMAGE.I_MENUBACKGROUND)!!,
            GameSystem.Companion.Instance.GetScreenScale()!!.x.toInt(),
            GameSystem.Companion.Instance.GetScreenScale()!!.y.toInt(),
            true
        )
        headerText = TextEntity.Companion.Create(
            GameSystem.Companion.Instance.ScreenScale!!.x * 0.5f,
            0f,
            "Scores",
            100f
        )
        headerText!!.SetPosY(headerText!!.GetHeight() * 0.5f)
        headerText!!.SetHeader(true)
        backButton = BackEntity.Companion.Create(
            0f,
            0f,
            headerText!!.height * 1.3f,
            headerText!!.height * 1.3f,
            "MainMenu"
        )
        backButton!!.SetPosX(backButton!!.width * 0.5f)
        backButton!!.SetPosY(backButton!!.height * 0.5f)

        HighScorePaint.color = Color.BLACK
        HighScorePaint.textSize = 50f

        Coinbmp = BitmapFactory.decodeResource(_view.resources, R.drawable.coin)
        moneyPaint.color = Color.YELLOW
        moneyPaint.alpha = 200
        moneyPaint.textSize = 50f
        moneyPaint.textAlign = Paint.Align.LEFT
        moneyPaint.isFakeBoldText = true
    }

    override fun OnExit() {
        EntityManager.Companion.Instance.EmptyEntityList()
    }

    override fun Render(_canvas: Canvas) {
        // Render anything
        _canvas.drawBitmap(background!!, 0f, 0f, null)
        ParticleManager.Companion.Instance.Render(_canvas)
        EntityManager.Companion.Instance.Render(_canvas)
        var count = 0
        for (i in 0..4) {
            if (i >= GameSystem.Companion.Instance.GetHighScores()!!.size) {
                _canvas.drawText(
                    (i + 1).toString() + ".---------",
                    GameSystem.Companion.Instance.ScreenScale!!.x * 0.1f,
                    GameSystem.Companion.Instance.GetScreenScale()!!.y * 0.2f + HighScoreTextBounds.height() * 1.2f * count,
                    HighScorePaint
                )
                val scoreText = "0"
                HighScorePaint.getTextBounds(scoreText, 0, scoreText.length, HighScoreTextBounds)
                _canvas.drawText(
                    scoreText,
                    GameSystem.Companion.Instance.ScreenScale!!.x * 0.9f - HighScoreTextBounds.width(),
                    GameSystem.Companion.Instance.GetScreenScale()!!.y * 0.2f + HighScoreTextBounds.height() * 1.2f * count,
                    HighScorePaint
                )
                ++count
                continue
            }


            _canvas.drawText(
                (i + 1).toString() + "." + GameSystem.Companion.Instance.GetHighScores()!!
                    .get(i).GetName(),
                GameSystem.Companion.Instance.ScreenScale!!.x * 0.1f,
                GameSystem.Companion.Instance.GetScreenScale()!!.y * 0.2f + HighScoreTextBounds.height() * 1.2f * count,
                HighScorePaint
            )
            val depthText = "" + GameSystem.Companion.Instance.GetHighScores()!!.get(i).GetScore()
            HighScorePaint.getTextBounds(depthText, 0, depthText.length, HighScoreTextBounds)
            _canvas.drawText(
                depthText,
                GameSystem.Companion.Instance.ScreenScale!!.x * 0.9f - HighScoreTextBounds.width(),
                GameSystem.Companion.Instance.GetScreenScale()!!.y * 0.2f + HighScoreTextBounds.height() * 1.2f * count,
                HighScorePaint
            )
            ++count
        }

        //Money
        val moneyText = "" + PlayerInfo.Companion.Instance.GetMoney()
        moneyPaint.getTextBounds(moneyText, 0, moneyText.length, moneyTextBounds)
        _canvas.drawBitmap(
            Coinbmp!!,
            Coinbmp!!.width * 0.5f,
            Coinbmp!!.height * 0.5f + headerText!!.height * 1.3f,
            null
        )
        _canvas.drawText(
            moneyText,
            Coinbmp!!.width + Coinbmp!!.width * 0.7f,
            Coinbmp!!.height + headerText!!.height * 1.3f + moneyTextBounds.height() * 0.5f,
            moneyPaint
        )
        //Money
    }

    override fun Update(_dt: Float) {
        if (TouchManager.Companion.Instance.HasTouch()) {
            for (i in 0..2) {
                ParticleManager.Companion.Instance.CreateParticle(
                    PlayerInfo.Companion.Instance.GetEffectType(),
                    TouchManager.Companion.Instance.GetPosX().toFloat(),
                    TouchManager.Companion.Instance.GetPosY().toFloat()
                )
            }
        }

        GameSystem.Companion.Instance.m_bubbleTimer += _dt
        if (GameSystem.Companion.Instance.m_bubbleTimer >= bubbleSpawnTime) {
            val newParticle: ParticleObject =
                ParticleManager.Companion.Instance.FetchParticle(PARTICLETYPE.P_BUBBLE)
            val diameter = (Math.random() * 50 + 15).toInt()
            newParticle.width = diameter.toFloat()
            newParticle.height = diameter.toFloat()
            newParticle.position.x = (Math.random()
                .toFloat() * (GameSystem.Companion.Instance.GetScreenScale()!!.x - newParticle.width)) + (newParticle.width * 0.5f)
            newParticle.position.y =
                GameSystem.Companion.Instance.GetScreenScale()!!.y + newParticle.height
            newParticle.velocity.x = Math.random().toFloat() * 40 - 20
            newParticle.velocity.y = -(Math.random().toFloat() * 160 + 80)
            newParticle.SetBMP(
                Bitmap.createScaledBitmap(
                    ImageManager.Companion.Instance.GetImage(
                        IMAGE.I_BUBBLE
                    )!!, diameter, diameter, true
                )
            )
            GameSystem.Companion.Instance.m_bubbleTimer = 0f
        }
        ParticleManager.Companion.Instance.Update(_dt)
        EntityManager.Companion.Instance.Update(_dt)
    }
}
